re-work miniperl build on Win32

tonycoz opened the pull request:
These commits re-work the miniperl build:
- previously the miniperl config.h was overwritten by the full perl config, which, if the objects for miniperl were rebuilt, could lead to the newly compiled miniperl objects not matching the existing objects.  This also caused a problem @khwilliamson encountered where a Perl_putenv() definition was confused about whether the build was threaded or not.  Using a separate config.h (now in mini/ for miniperl) avoids this confusion.
- previously the miniperl object files didn't depend on their corresponding source files, presumably with the intent to avoid the above problem.  This could be frustrating if the developer was testing core changes by rapidly iterating on miniperl.  Since the miniperl object are now generated from a distinct config.h, this has been switched to a pattern (GNU make) or implicit rule (nmake) which adds an implicit dependency on the source file.

I believe this fixes some of the problems in #20749

tonycoz committed:
win32/GNUmakefile: move miniperl's config.h to mini\

On Win32 the objects for miniperl can differ from those used for the
main perl executable/DLL, typically without threads and without
USE_IMP_SYS.

This involved producing a dummy config.h used to build the miniperl
objects, which was then later replaced with the config.h for the full
build.

This could cause other problems during rebuilds, since code that
assumed PERL_IS_MINIPERL meant no threads could fail to build, and
if the miniperl objects needed to be rebuilt for some reason, they'd
be rebuilt with the full config.h, possibly mismatching the rest of
the objects.

tonycoz committed:
win32/GNUmakefile: use pattern rules to build the miniperl objects

This adds implicit dependencies between the objects and their sources
so that touching the .c file and rebuilding rebuilds the miniperl
object and executable.

tonycoz committed:
win32/Makefile: move the miniperl config.h to mini\

tonycoz committed:
win32/Makefile: build miniperl objects with implicit rules

This adds implicit dependencies of the mini/ objects on their
source files, so they (and miniperl itself) are rebuilt if their
source updates.

demerphq commented on win32/GNUmakefile:
```diff
@@ -1420,23 +1420,27 @@ else
 #separate line since this is sentinal that this target is done
 	@rem. > $(MINIDIR)\.exists
 
+$(MINIDIR)\\%$(o): %.c
```
is the double backslash required here? just above it doesnt seem to be required.  Would a comment be in order?

Anyway, other than that LGTM.

tonycoz replied:
A single `\` quotes the `%`, so an extra `\` is needed to quote the `\`.

https://www.gnu.org/software/make/manual/html_node/Static-Usage.html#index-_005c-_0028backslash_0029_002c-to-quote-_0025-1

(edit: better reference)

tonycoz commented:
Draft since I want to do some further checks.  Mostly I'm worried about `#include "config.h"` for miniperl checking the current directory first.

tonycoz committed:
win32/GNUmakefile, win32/Makefile: define a macro for the mini/config.h

tonycoz committed:
win32/@makefiles: move the full perl config.h to full/

Unfortunately compilers will look in . for #include "..." before any
-I specified, this meant that objects for miniperl.exe could be
built with the full perl configuration which doesn't match the
configuration used to build the miniperl objects the first time
around.

config.h is referenced by a few other tools, so those also needed to
be updated to find the moved config.h.

tonycoz committed:
win32/@makefiles: only update config.sh if it changes

But also ensure that the config.sh mod time matches the latest of its
dependencies so that it isn't rebuilt the next time around.

For GNU make, this prevents anything that depends on config.sh
being unnecessarily rebuilt.

nmake uses the current time as the modification time, so this doesn't
prevent an unneed rebuild for dependencies for the current build, but
does for the next build.

tonycoz committed:
win32/@makefiles: only update "Extensions_static" if needed

Where needed is when: the file changes or any of the static
libraries are newer than Extensions_static

Previously it would always be updated, so perl5xx.dll would always
be rebuilt.

With gmake this prevents perl5xx.dll from being rebuilt.

For nmake it unfortunately doesn't, but may prevent updates in later
rebuilds.

tonycoz commented:
@khwilliamson could you please see if this solves any of your problems from #20749

khwilliamson commented:
Yes it works AFAICT

tonycoz commented:
Applied manually.
